toolbox/src/base/RobotInterface.cpp

```diff
@@ -0,0 +1,447 @@
+    // ----------------------------
+
+    // Initialize RF
+    // Workaround for the fact that ResourceFinder initializes the network by itself. See YARP#1014
+    using namespace yarp::os;
+    Network network;
+    ResourceFinder& rf = ResourceFinder::getResourceFinderSingleton();
+    rf.configure(0, 0);
+
+    // Get the absolute path of the urdf file
+    std::string urdf_file = getConfiguration().getUrdfFile();
+    std::string urdf_file_path = rf.findFile(urdf_file.c_str());
+
+    // Load the reduced model into KinDynComputations
+    // ----------------------------------------------
+
+    // Load the joint list
+    std::vector<std::string> controlledJoints = getConfiguration().getControlledJoints();
+
+    // Use ModelLoader to load the reduced model
+    iDynTree::ModelLoader mdlLoader;
+    if (!mdlLoader.loadReducedModelFromFile(urdf_file_path, controlledJoints)) {
+        Log::getSingleton().error("ToolboxSingleton: impossible to load " + urdf_file + ".");
+        Log::getSingleton().error("Probably the joint list contains an entry not present in the urdf model.");
+        return false;
+    }
+
+    // Add the loaded model to the KinDynComputations object
+    return m_kinDynComp->loadRobotModel(mdlLoader.model());
+}
+
+bool RobotInterface::initializeRemoteControlBoardRemapper()
+{
+    // Setup the RemoteControlBoardRemapper
+    yarp::os::Property options;
+    options.put("device", "remotecontrolboardremapper");
+    options.put("axesNames", vectorToString(m_config.getControlledJoints()));
+    std::string prefix = "/" + m_config.getRobotName() + "/";
+    options.put("remoteControlBoards", vectorToString(m_config.getControlBoardsNames(), prefix));
+    options.put("localPortPrefix", m_config.getLocalName());
+    yarp::os::Property& remoteCBOpts = options.addGroup("REMOTE_CONTROLBOARD_OPTIONS");
+    remoteCBOpts.put("writeStrict","on");
+
+    // Allocate the device driver
+    m_robotDevice = std::make_shared<yarp::dev::PolyDriver>();
+
+    if (!m_robotDevice) {
+        return false;
+    }
+
+    // Open the device driver
+    if (m_robotDevice->open(options) && m_robotDevice->isValid()) {
+        return true;
+    }
+
+    // Remove garbage if the opening fails
+    m_robotDevice.reset();
+    return false;
+}
+
+// OTHER METHODS
+// =============
+
+template <typename T>
+std::weak_ptr<T> RobotInterface::getDevice(std::shared_ptr<T> device)
+{
+    if (!device) {
+        // Blocks which require the RemoteControlBoardRemapper need to retain / release it
+        // in their initialization / terminate phase;
+        assert(m_robotDevice);
+        if (!m_robotDevice) {
+            // Return an empty weak pointer
+            return std::weak_ptr<T>();
+        }
+
+        T* ptr = nullptr;
+        if (!m_robotDevice->view(ptr)) {
+            // Return an empty weak_ptr
+            return std::weak_ptr<T>();
+        }
+        // Store ptr into the smart pointer
+        device.reset(ptr);
```

Comment:
This line is transferring ownership of the object pointer by `ptr` to the shared_ptr, but the ownership of the object (i.e. who will destroy it) is still maintained by the PolyDriver object, and there is no way to share it, and it is not implemented using smart pointers. I am surprised that this is not causing any kind of "double free" problems.

Reply:
@diegoferigo @francesco-romano

Reply:
@traversaro I was not aware of this, and maybe I did not completely got your point here. This (templated) method returns a weak pointer, and it should not deallocate the object in its destructor, am I wrong?

Reply:
> A solution I can see is using weak_ptr also for the internal members. Thanks a lot for this comment, I was not aware of the implementation of the device.

For having a `weak_ptr` a shared_ptr must have the ownership of the object, and this is not the case. The only solution I can think of is using raw pointers.

Reply:
Yes, no other clean solution exist unfortunately. We'll live well anyways :smile: